src/cryptonote_core/service_node_swarm.cpp

```diff
@@ -0,0 +1,270 @@
+    LOG_PRINT_L2("Calculated threshold: " << threshold);
+    return threshold;
+  };
+
+  prod_static void create_new_swarm_from_excess(swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt)
+  {
+    while (calc_excess(swarm_to_snodes) >= calc_threshold(swarm_to_snodes))
+    {
+      LOG_PRINT_L2("New swarm creation");
+      std::vector<crypto::public_key> new_swarm_snodes;
+      std::vector<excess_pool_snode> excess_pool;
+
+      for (const auto &entry : swarm_to_snodes)
+      {
+        if (entry.second.size() > EXCESS_BASE)
+        {
+          for (const auto &sn_pk : entry.second)
+          {
+            excess_pool.push_back({sn_pk, entry.first});
+          }
+        }
+      }
+
+      while (new_swarm_snodes.size() < NEW_SWARM_SIZE)
+      {
+        const auto idx = uniform_distribution_portable(mt, excess_pool.size());
+        auto &random_excess = excess_pool.at(idx);
+        const auto random_sn_pk = random_excess.public_key;
+        const auto random_sn_swarm_id = random_excess.swarm_id;
+        LOG_PRINT_L2("Taking from swarm : " << random_sn_swarm_id);
+        /// Remove service node from swarm
+        auto &swarm_sn_vec = swarm_to_snodes.at(random_sn_swarm_id);
+        swarm_sn_vec.erase(std::remove(swarm_sn_vec.begin(), swarm_sn_vec.end(), random_sn_pk), swarm_sn_vec.end());
+        if (swarm_sn_vec.size() <= EXCESS_BASE)
+        {
+          /// Remove the swarm's snodes from the pool
+          excess_pool.erase(std::remove_if(excess_pool.begin(),
+                                          excess_pool.end(),
+                                          [&](const excess_pool_snode &excess) {
+                                            return excess.swarm_id == random_sn_swarm_id;
+                                          }),
+                            excess_pool.end());
+        }
+        else
+        {
+          /// Remove only the one snode from the pool
+          excess_pool.erase(excess_pool.begin() + idx);
+        }
+
+        /// Add to new swarm
+        new_swarm_snodes.push_back(random_sn_pk);
+      }
+      const auto new_swarm_id = get_new_swarm_id(mt, swarm_to_snodes);
+      swarm_to_snodes.insert({new_swarm_id, std::move(new_swarm_snodes)});
+      LOG_PRINT_L2("Created new swarm from excess: " << new_swarm_id);
+    }
+  }
+
+  prod_static void calc_swarm_sizes(const swarm_snode_map_t &swarm_to_snodes, std::vector<swarm_size> &sorted_swarm_sizes)
+  {
+    sorted_swarm_sizes.clear();
+    for (const auto &entry : swarm_to_snodes)
+    {
+      sorted_swarm_sizes.push_back({entry.first, entry.second.size()});
+    }
+    std::sort(sorted_swarm_sizes.begin(),
+              sorted_swarm_sizes.end(),
+              [](const swarm_size &a, const swarm_size &b) {
+                return a.size < b.size;
+              });
+  }
+
+  prod_static void assign_snodes(const std::vector<crypto::public_key> &snode_pubkeys, swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt)
```

Comment:
Would be good a have a description/comment of how this function works/ what it does.

Reply:
Added